docs/core/diagnostics/metrics-collection.md

```diff
@@ -0,0 +1,404 @@
+connect it to a Prometheus data source.
+
+2. Create a Grafana dashboard by clicking the + icon on the left toolbar in the Grafana web portal, then select 'Dashboard'. In dashboard
+editor that appears enter 'Hats Sold/Sec' as the Title and 'rate(hats_sold[5m])' in the PromQL expression field. It should look like this:
+[hats-sold-dashboard-editor](media\grafana-hats-sold-dashboard-editor.png)
+
+3. Click Apply to save and view our simple new dashboard.
+[hats-sold-dashboard](media\grafana-hats-sold-dashboard)
+
+
+### Creating a custom collection tool using the .NET <xref:System.Diagnostics.Metrics.MeterListener> API
+
+The .NET <xref:System.Diagnostics.Metrics.MeterListener> API allows creating custom in-process logic to observe the measurements
+being recorded by <xref:System.Diagnostics.Metrics.Meter?displayProperty=nameWithType>. For guidance creating custom
+logic compatible with the older EventCounters instrumentation see [here](event-counters.md).
+
+Modify the code of Program.cs to use the MeterListener like this:
+```C#
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Threading;
+
+class Program
+{
+    static Meter s_meter = new Meter("HatCo.HatStore", "1.0.0");
+    static Counter<int> s_hatsSold = s_meter.CreateCounter<int>(name: "hats-sold", 
+                                                                unit: "Hats",
+                                                                description: "The number of hats sold in our store"); 
+
+    static void Main(string[] args)
+    {
+        MeterListener meterListener = new MeterListener();
+        meterListener.InstrumentPublished = (instrument, listener) =>
+        {
+            if(instrument.Meter.Name == "HatCo.HatStore")
+            {
+                listener.EnableMeasurementEvents(instrument);
+            }
+        };
+        meterListener.SetMeasurementEventCallback<int>(OnMeasurementRecorded);
+        meterListener.Start();
+
+        Console.WriteLine("Press any key to exit");
+        while(!Console.KeyAvailable)
+        {
+            // Pretend our store has a transaction each second that sells 4 hats
+            Thread.Sleep(1000);
+            s_hatsSold.Add(4);
+        }
+
+        meterListener.Dispose();
+    }
+
+    static void OnMeasurementRecorded<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string,object>> tags, object? state)
+    {
+        Console.WriteLine($"{instrument.Name} recorded measurement {measurement}");
+    }
+}
+```
+
+When run the application now runs our custom callback on each measurement:
+```dotnetcli
+> dotnet run
+Press any key to exit
+hats-sold recorded measurement 4
+hats-sold recorded measurement 4
+hats-sold recorded measurement 4
+hats-sold recorded measurement 4
+...
+```
+
+Lets break down what happens in the example above. 
+```C#
+using var meterListener = new MeterListener();
```

Comment:
>using [](http://example.com/codeflow?start=0&length=5)

interesting, you already using it here :-)

Reply:
Yeah I was switching back and forth between the two approaches trying to get a better flow for the explanatory text. I wound up with inconsistent text : ) It should be consistent now as `using MeterListener meterListener = ...`